src/main/scala/exercise4/autozion/package.scala

```diff
@@ -0,0 +1,97 @@
+package exercise4
+
+import zio.blocking.{Blocking, effectBlocking}
+import zio.clock.Clock
+import zio.console._
+import zio.duration.{Duration, durationLong}
+import zio.random.{Random, nextIntBounded, nextLongBetween, nextPrintableChar, nextString, nextUUID}
+import zio.{Has, RIO, Schedule, ZIO}
+import java.io.IOException
+
+
+package object autozion {
+  type MyEnv = Has[News] with Has[JobBoard] with Has[CompletedJobsHub] with Random with Clock with Blocking with Console
+  type ElderEnv = Has[JobBoard] with Random with Clock
+  type WorkerEnv = Has[JobBoard] with Has[CompletedJobsHub] with Clock with Blocking  with Console with Random
+  type OverseerEnv = Has[News] with Has[CompletedJobsHub] with Clock
+  type PraiserEnv = Has[News] with Has[CompletedJobsHub] with Clock
+  type ReporterEnv = Has[News] with Clock with Console
+
+  sealed trait Job
+
+  trait Robot {
+    def name: String
+
+    def work: RIO[MyEnv, Unit]
+  }
+
+  case class PendingJob(name: String, duration: Duration) extends Job
+
+  case class CompletedJob(name: String, completedBy: Robot) extends Job
+
+  case class Elder(name: String = "Elder") extends Robot {
+
+    override def work: RIO[ElderEnv, Unit] = (for {
+      randomLong <- nextLongBetween(2, 5)
+      duration = randomLong.seconds
+      jobID <- nextUUID
+      jobName = s"do $jobID for ${duration.toSeconds.toString} seconds from $name"
+      job = PendingJob(jobName, duration)
+      _ <- JobBoard.submit(job)
+    } yield ()).repeat(Schedule.spaced(2.seconds).delayed(_ => 5.seconds)).unit
+  }
+
+  case class Worker(name: String = "Worker") extends Robot {
+
+    override def work: RIO[WorkerEnv, Unit] = (for {
```

Comment:
As already noted, solution accepted. What I had in mind was something along the lines of (albeit not so generic):

```scala
def withReboot[R, E <: Throwable, A](work: ZIO[R, E, A]): URIO[R with Clock, A] = 
work.retry(Schedule.spaced(2.seconds)).orDie
```
This just takes an effect (work) and retries it infinitely if it fails, with a 2 second delay between each retry. `orDie` is just used to erase the (actually impossible since infinite retires) error type

It can then be used in the following way:

```scala
_ <- withReboot(Worker("W").work()).fork
```

and the work method, which fails with a 1 in 5 chance
```scala
  override def work(): ZIO[Has[CompletedJobsHub] with Clock with Has[JobBoard] with Random, RuntimeException, Nothing] =
    (for {
      job <- JobBoard.take()
      randInt <- nextIntBetween(0, 5)
      _ <- ZIO.when(randInt == 0)(ZIO.fail(new RuntimeException("ERROR, ERROR!")).ensuring(JobBoard.submit(job)))
      _ <- ZIO.sleep(job.duration)
      _ <- CompletedJobsHub.publish(CompletedJob(job.name, this))
    } yield ()).forever
}
```